File: JavaScript/Easy/symmetricTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

// TODO: find a way to do it <1 ms and <58MB memory space thru dynamic programming.
var isSymmetric = function (root) {
  let returnValue = true;

  const innerTraverse = (left, right) => {
    if (left == null && right == null) return;

    if ((left == null && right) || (left && right == null)) {
      returnValue = false;
      return;
    }

    if (!(left.val == right.val)) {
      returnValue = false;
      return;
    }

    innerTraverse(left.left, right.right);

    if (!returnValue) {
      return;
    }

    innerTraverse(left.right, right.left);
  };

  innerTraverse(root.left, root.right);
  return returnValue;
};

class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

const parent = new TreeNode(2);

const leftChild = new TreeNode(3);
const rightChild = new TreeNode(3);

const leftChildLeft = new TreeNode(4);
const leftChildRight = new TreeNode(5);

const rightChildRight = new TreeNode(4);

parent.left = leftChild;
parent.right = rightChild;

leftChild.left = leftChildLeft;
leftChild.right = leftChildRight;

rightChild.right = rightChildRight;

console.log(isSymmetric(parent));

File: Cpp/Easy/SymmetricTree.cpp
#include <iostream>

using std::cout;

struct TreeNode {
	TreeNode* left;
	TreeNode* right;
	int val;

    TreeNode() : val(0), left(nullptr), right(nullptr) {};
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {};
	TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

bool isSymmetric(TreeNode* root);
void innerTraverse(TreeNode* left, TreeNode* right, bool& retVal);


int main() {
    TreeNode left_child_left(3);
    TreeNode left_child_right(4);

    TreeNode right_child_left(4);
    TreeNode right_child_right(3);

    TreeNode left_child(2, &left_child_left, &left_child_right);
    TreeNode right_child(2, &right_child_left, &right_child_right);

    TreeNode parent(1, &left_child, &right_child);

    std::cout << isSymmetric(&parent);

    return 0;
}

bool isSymmetric(TreeNode* root) {
    bool returnValue = false;

    innerTraverse(root->left, root->right, returnValue);

    return returnValue;

	return false;
}

void innerTraverse(TreeNode* left, TreeNode* right, bool& retVal) {
    if (left == nullptr && right == nullptr)
        return;

    if (left == nullptr || right == nullptr) {
        retVal = false;
        return;
    }

    if (left->val != right->val) {
        retVal = false;
        return;
    }
        
    innerTraverse(left->left, right->right, retVal);

    if (!retVal)
        return;

    innerTraverse(left->right, right->left, retVal);

    return;
}